sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -236,6 +271,138 @@ private Session session() {
+        //String targetIndexName = ".resized" + sourceIndexName + String.valueOf(System.currentTimeMillis());
+
+        int targetNumberOfShards = analysis.tableParameter().settings().getAsInt(SETTING_NUMBER_OF_SHARDS, 0);
+        validateForResize(sourceIndexName, targetNumberOfShards);
+
+        List<ChainableAction<Long>> actions = new ArrayList<>(7);
+
+        // 0. pickup node
+        DiscoveryNode resizeNode = getNodeForResize();
+
+        // 1. change allocation
+        actions.add(new ChainableAction<>(
+            () -> restrictAllocationToNode(resizeNode.getName(), sourceIndexName),
+            () -> revertAllocation(sourceIndexName)
+        ));
+
+        // todo ensure green state
+
+        // 2. invoke shrink operation
+        actions.add(new ChainableAction<>(
+            () -> shrinkIndex(targetIndexName, sourceIndexName, targetNumberOfShards),
+            () -> dropIndex(RelationName.fromIndexName(targetIndexName), isPartitioned)
+        ));
+
+        // 3. atomic index name exchange
+        actions.add(new ChainableAction<>(
+            () -> closeTable(sourceIndexName, targetIndexName),
+            () -> openTable(sourceIndexName, targetIndexName)
+        ));
+        actions.add(new ChainableAction<>(
+            () -> exchangeTableNames(RelationName.fromIndexName(targetIndexName), RelationName.fromIndexName(sourceIndexName), isPartitioned),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+        actions.add(new ChainableAction<>(
+            () -> openTable(sourceIndexName),
+            () -> closeTable(sourceIndexName)
+        ));
+
+        // 4. drop original index
+        actions.add(new ChainableAction<>(
+            () -> dropIndex(RelationName.fromIndexName(targetIndexName), isPartitioned),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // 5. revert index allocation
+        actions.add(new ChainableAction<>(
+            () -> revertAllocation(sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+        return ChainableActions.run(actions);
+    }
+
+    private void validateForResize(String indexName, int targetNumberOfShards) {
+        final int currentNumberOfShards = clusterService.state().metaData().index(indexName).getNumberOfShards();
+        if (currentNumberOfShards == targetNumberOfShards) {
+            throw new IllegalArgumentException(
+                String.format("Table/partition is already allocated <%d> shards", currentNumberOfShards));
+        } else if (targetNumberOfShards > currentNumberOfShards){
+            throw new IllegalArgumentException("Increasing the number of shards is not supported");
+        }
+
+        final Boolean blocks_write = clusterService.state().metaData().index(indexName)
+            .getSettings().getAsBoolean(INDEX_BLOCKS_WRITE_SETTING.getKey(), false);
```

Comment:
let's discuss this as in the ES documentation it is mentioned that in order to prepare index for shrinking you need to mark it as read only: `"index.blocks.write": true`

Reply:
@mfussenegger yes, it will fail later. However does't that mean that we have already performed the shard allocation which might be an expensive operation and that in addition we have to undo it? If this holds, I would say that it makes sense to bail out before any of this occurs if we know that it will fail on a later stage anyway.

Reply:
decided to keep all checks that we know they can cause the shrink operation to fail